src/beamlime/applications/handlers.py

```diff
@@ -0,0 +1,115 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+import pathlib
+from dataclasses import dataclass
+from typing import Any, NewType
+
+import plopp as pp
+import scipp as sc
+
+from beamlime.logging import BeamlimeLogger
+
+from ._workflow import Events, FirstPulseTime, Histogrammed, WorkflowPipeline
+from .base import HandlerInterface, MessageProtocol
+
+
+@dataclass
+class HistogramUpdated(MessageProtocol):
+    content: pp.graphics.basefig.BaseFig
+    sender: type
+    receiver: type
+
+
+@dataclass
+class RawDataSent(MessageProtocol):
+    content: Events
+    sender: type
+    receiver: type
+
+
+class DataReductionHandler(HandlerInterface):
+    """Data reduction handler to process the raw data and update the histogram.
+
+    It receives a list of events, and reduces them into a histogram.
+    It also triggers the update of a plot stream node.
+    """
+
+    def __init__(self, pipeline: WorkflowPipeline) -> None:
+        self.output_da: Histogrammed
+        self.stream_node: pp.Node
+        self.pipeline = pipeline
+        super().__init__()
+
+    def format_received(self, data: Any) -> str:
+        return f"{len(data)} pieces of {Events.__name__}"
+
+    def process_first_input(self, da: Events) -> None:
+        first_pulse_time = da[0].coords['event_time_zero'][0]
+        self.pipeline[FirstPulseTime] = first_pulse_time
+
+    def process_first_output(self, data: Histogrammed) -> None:
+        self.output_da = Histogrammed(sc.zeros_like(data))
+        self.info("First data as a seed of histogram: %s", self.output_da)
+        self.stream_node = pp.Node(self.output_da)
+        self.figure = pp.figure1d(
+            self.stream_node,
+            title="Wavelength Histogram",
+            grid=True,
+        )
```

Comment:
Can we move the plotting code out of the data-reduction handler? Instead, put it into the `PlotSaver`? This coupling of the processing with the visualization adds coupling.